# FEATURE - Signer Pool Manager

blackironj opened the pull request:
### 추가 및 수정
- ErrrorMsgType 추가

- Signer Protobuf 수정
  - Net-plugin에서 Signer 와 ECDH Key exchange를 하므로, Signer로부터 MSG_JOIN, RESPONSE1, SUCCESS를 받았을때, 다른 plugin으로 전달이 필요없습니다.
  - 해당 메시지를 받았을 때, 해당 RPC 를 통해 즉각적으로 답하는 형태가 맞다고 생각하여 수정하였습니다.

- SignerPoolManager
  - SignerPool 추가
  - ECDH exchange 관련 메시지 처리 (JOIN / RESPONSE1 /SUCCESS)

- MessageBuilder 
  - MSG_LEAVE 제거 ( NetPlugin에서 signer pool 관리로 해당 메시지를 발생시킬 필요가 없어짐)
  - CHALLENGE / RESPONSE2  / ACCEPT 추가

- MessageHandler
  - JOIN / RESPONSE1 / SUCCESS 처리

- ~MessageParser~
  - ~RequesterType에 따라 parsing 변경~

- MessagePacker
  - Mac algorithm type에 따라 packMessage 하도록 수정

- Signer로부터 온 메시지 Parsing 시 HMAC 검증하는 코드 추가
- Signer에게 보낼 메시지에 HMAC을 붙이도록 함.

### 기타 수정
- Signer conn manager -> signer conn table 파일 이름 수정

Commit:
# EDIT - Signer protobuf

- ECDH Key exchange 가 net-plugin 단에서 처리하기 때문에 protobuf 수정

Commit:
# ADD - Signer pool manager

- Signer Pool Manager
  - Singer pool 추가
  - ECDH exchange 관련 message 처리 ( JOIN / RESPONSE1 / SUCCESS )

- Message builder
  - CHALLENGE / RESPONSE2 / ACCEPT

- signer conn manager -> signer conn table 파일 이름 수정
- message.hpp에 ErrMsgType 추가
- network_config에 임시로 사용할 MERGER_ID_b58 추가

Commit:
# EDIT - RPC - service

- packMsg / makeHeader 를 message_packer.hpp 로 이동
- Requester Type (Signer / Merger ) 따라 MessageParsing 구분
- MessageHandler
  - JOIN / RESPONSE1 / SUCCESS 를 처리할 수 있도록 함.

- MessageParser 에서 Signer msg parsing 할때 HMAC 검증 코드 추가
- 메시지를 보낼때  특정 signer message(SSIG ,ACCEPT) 일 경우 HMAC을
붙이도록 함

- 기타 : SignerPool 에서 사용하지 않는 std::mutex 제거

blackironj commented:
@park53kr 
놓친부분 하나 있네요. 수정해서 다시 올릴게요. (Signer Message parsing 부분)

morishjs commented on src/plugins/net_plugin/include/signer_pool_manager.hpp:
```diff
@@ -0,0 +1,201 @@
+    case MessageType::MSG_JOIN: {
+      if (!isJoinable()) {
+        logger::ERROR("[ECDH] Signer pool is full");
+        throw ErrorMsgType::ECDH_MAX_SIGNER_POOL;
```
throw를 던지고 그걸 처리하는 코드가 없네요
이런건 단순히 std::exception 으로 처리하는 것보다 exception class 새로 만들어서 처리하는게 좋아보입니다.

blackironj replied:
지금 `rpc_services.cpp` 에서  SignerService::proceed() 에서 이런식으로 처리 하고 있긴있습니다. 
```
 catch (exception &e) { 
      m_reply.set_status(Reply_Status_INTERNAL);
    } catch (ErrorMsgType err_msg_type) {
      m_reply.set_status((Reply_Status)err_msg_type);
    }
```
이부분 exception class 만들어서 하는게 좋을까요?

morishjs replied:
아.. 있었군요 따로 안만들어도 될 것 같습니다.

morishjs commented on src/plugins/net_plugin/net_plugin.cpp:
```diff
@@ -283,9 +285,11 @@ class NetPluginImpl {
   }
 
   void sendMessage(OutNetMsg &out_msg) {
+
```
가끔 @blackironj  코드를 보다보면 함수 정의 첫번째줄에 line-break가 있을때도 있고 없을때도 있는데 특별한 이유가 없다면 line-break하지 않도록 해주세요

blackironj replied:
주의하도록하고 수정하겠습니다.

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -35,12 +41,21 @@ class MessageParser {
-      vector<uint8_t> hmac(packed_msg.begin() + HEADER_LENGTH + body_size, packed_msg.end());
-
-      // TODO : verify HMAC
+    if constexpr (T == RequesterType::SIGNER) {
```
MessageParser가 하는 역할은 단순히 packed_msg 를 파싱해서 InNetMsg를 리턴하는 것입니다. 그런데 HMAC 의 존재여부까지 검사하게 되면서 코드가 너무 장황해졌는데요. 예를 들면, MessageParser를 생성할때 signer_pool_manager를 무조건 넘겨야 한다든지 RequesterType을 넘겨줘야 하는 문제들이요. 함수 인자를 최대한 많이 안넘겨주는게 가장 best case이기 때문에 함수가 인자를 2개 초과(템플릿 인자 포함) 해서 넘겨주는 것으로 변경되었을때 진지하게 고민해봐야 돼요.

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -35,12 +41,21 @@ class MessageParser {
-
-      // TODO : verify HMAC
+    if constexpr (T == RequesterType::SIGNER) {
+      if (msg_header->mac_algo_type == MACAlgorithmType::HMAC) {
```
이 조건에 있는 코드들은 함수로 하나 빼서 따로 호출하는게 맞는것 같습니다. 210번째 줄 밑이나 적절한 곳에서 함수 호출하면 될 것 같습니다.

blackironj replied:
수정하도록 하겠습니다.

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -174,27 +201,37 @@ void SignerService::proceed() {
+          auto reply_packed_msg = MessagePacker::packMessage(reply_msg.value());
+          if (reply_msg.value().type == MessageType::MSG_ACCEPT) {
+            auto receiver_id_b58 = input_data.value().sender_id;
+            auto hmac_key = m_signer_pool_manager->getHmacKey(receiver_id_b58);
```
`m_signer_pool_manager->getHmacKey(receiver_id_b58)`

없을 수도 있는거 아닌가요?

morishjs replied:
```
if (reply_msg.value().type == MessageType::MSG_ACCEPT) {
...
}
```
코드도 결국엔 packMessage의 과정 중 하나라서 packMessage에 넣는게 맞는 것 같습니다.

blackironj replied:
1. 있는지 확인해야 할거 같네요. 수정할게요.
2. 이 부분도 수정하겠습니다.

Commit:
# EDIT - Message Parser, Message Packer

- MessagePacker에서 packMessage를 MAC algorithm 타입에 따라서 하도록
수정
- MessageParser::parserMessage에서 HMAC 검증하는 부분을 SignerService
객체로 이동

blackironj commented:
`SignerService::proceed()`  쪽 코드 부분이 가독성이 너무 떨어지네요. 차후 리팩토링하거나 수정하겠습니다.
